vmm/src/config.rs

```diff
@@ -3333,11 +3399,164 @@ mod tests {
         assert!(still_valid_config.validate().is_ok());
 
-        let mut invalid_config = valid_config;
+        let mut invalid_config = valid_config.clone();
         invalid_config.platform = Some(PlatformConfig {
             num_pci_segments: 16,
             iommu_segments: Some(vec![17, 18]),
         });
         assert!(invalid_config.validate().is_err());
+
+        let mut still_valid_config = valid_config.clone();
+        still_valid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        still_valid_config.disks = Some(vec![DiskConfig {
+            iommu: true,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(still_valid_config.validate().is_ok());
+
+        let mut still_valid_config = valid_config.clone();
+        still_valid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        still_valid_config.net = Some(vec![NetConfig {
+            iommu: true,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(still_valid_config.validate().is_ok());
+
+        let mut still_valid_config = valid_config.clone();
+        still_valid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        still_valid_config.pmem = Some(vec![PmemConfig {
+            iommu: true,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(still_valid_config.validate().is_ok());
+
+        let mut still_valid_config = valid_config.clone();
+        still_valid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        still_valid_config.devices = Some(vec![DeviceConfig {
+            iommu: true,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(still_valid_config.validate().is_ok());
+
+        let mut still_valid_config = valid_config.clone();
+        still_valid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        still_valid_config.vsock = Some(VsockConfig {
+            iommu: true,
+            pci_segment: 1,
+            ..Default::default()
+        });
+        assert!(still_valid_config.validate().is_ok());
+
+        let mut invalid_config = valid_config.clone();
+        invalid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        invalid_config.disks = Some(vec![DiskConfig {
+            iommu: false,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(invalid_config.validate().is_err());
```

Comment:
By adding `#[derive(PartialEq)]` to `ValidationError`, we can precisely assert on the error we are expecting, e.g. `assert_eq!(invalid_config.validate(), Err(ValidationError::OnIommuSegment(1)))`, which would be better?

Reply:
Probably worth doing in a follow-up PR. Since this whole test could be improved.